### PR TITLE
[Flink-21738][table]use cache in module to reduce unnecessary method calls

### PR DESCRIPTION
**Pull request by zoucao**
## What is the purpose of the change
use cache to reduce unnecessary method calls in `ModuleManager`, the modules involved are 'core' and 'hive'.

## Brief change log
- use cache to reduce unnecessary method calls

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluser with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zoucao** [comment]
cc @wuchong, please have a look, if have a time.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java [hunk 1]
```diff
@@ -173,16 +174,21 @@ public void useModules(String... names) {
...
+                                    .anyMatch(name::equalsIgnoreCase)) {
+                                moduleName.set(n);
+                                return true;
+                            }
+                            return false;
+                        });
+
+        if (moduleName.get() != null) {
+            LOG.debug("Got FunctionDefinition '{}' from '{}' module.", name, moduleName.get());
+            return loadedModules.get(moduleName.get()).getFunctionDefinition(name);
```
Comment (wuchong) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/module/ModuleManager.java:191]: We don't need use `AtomicReference` here, you can refactor this logic into while loop and return once the current module contains the requried function name.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/CoreModule.java [hunk 2]
```diff
@@ -29,19 +32,28 @@
 /** Module of default core metadata in Flink. */
 public class CoreModule implements Module {
     public static final CoreModule INSTANCE = new CoreModule();
+    private List<BuiltInFunctionDefinition> builtInFunctionDefinitions;
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/CoreModule.java:35]: can be `final`.

Comment (zoucao) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/CoreModule.java]: Thanks for your review，I have revised the above two.

**Review by wuchong — APPROVED** [review]
LGTM.

I appended a commit to improve the variable name a bit.

**Comment by zoucao** [comment]
Thanks for your help about this PR, jack! I will pay attention to it in the future.

**Pull request closed without merging**